### DISCUSSION TITLE
EVE Commands cannot be called from C++ code

### DISCUSSION THREAD

**Discussion by jwestes** [RudolphRiedel/FT800-FT813#64] [Ideas]
I would like to do away with tft.h and tft.c and use a Display class instead.  In order to do that, I had to wrap all the definitions in eve_commands.h with an extern "c" statement.

**Comment by Brechtjeu** [comment]
Hi, I have a similar intrest in including this in a cpp project, any way @jwestes, you can help me out on how you did it?

I did not seem to be successfull by only wrapping eve_commands.h in extern c.

I'm looking forward to any tips on how to include this in a cpp project

**Reply by jwestes** [reply]
That was the only change I had to make to use this library in cpp.  What sorts of problems are you having?

**Comment by RudolphRrr** [comment]
I am  not sure what the actual issue is since I am not using C++.
I am thinking about using C++ but I am not quite there, yet.

The first way I made my library work with Arduino, which is compiled in C++, was to just rename the .c files, so EVE_commands.c to EVE_commands.cpp.
But that was a pain to work with.

My current solution is to make the C++ callable from C and so I added EVE_cpp_wrapper.cpp.
So it is fine to have tft.c in an Arduino project.

But while using a .c file is fine for me you want/need to do it in C++ and not call TFT_init() but rather TFT.init()?

Ok, I just tried something, I added the extern "C" on top of EVE.h:

...
#ifndef EVE_H
#define EVE_H

#ifdef __cplusplus
extern "C" {
#endif

#pragma once

#include "EVE_target.h"
#include "EVE_config.h"
#include "EVE_commands.h"
...

And of course the closing bracket on the bottom:
...
#endif /*  EVE_GEN > 3 */

#ifdef __cplusplus
}
#endif

#endif /* EVE_H */

And EVE.h includes EVE_target.h, EVE_config.h and EVE_commands.h.

I can now put functions from EVE_commands.c into the Arduino setup() and loop()
and at least it compiles just fine.

If this is it, I will add it after some more testing.

**Comment by jwestes** [comment]
"But while using a .c file is fine for me you want/need to do it in C++ and not call TFT_init() but rather TFT.init()?"
Essentially, yes.  I created what I call a Display class and used the code in TFT_init() as a guide in my Init faction on that class.

I tried that same change and I get all sorts of "143:6 error: conflicting declaration of C function XXX" errors.  The only way I could get it to work is by not wrapping target.h in an extern "C".  So the top of my eve.h file looks like this.

#ifndef EVE_H
#define EVE_H

#pragma once

#include "EVE_target.h"

#ifdef __cplusplus
extern "C" {
#endif

#include "EVE_config.h"
#include "EVE_commands.h"

#ifdef __cplusplus
}
#endif

**Reply by RudolphRrr** [reply]
For what target and have you tried this more recently?

**Reply by jwestes** [reply]
I pulled the code from your repo as of 2/24/2023 and I just tried the change 5 minutes ago.  I'm targeting Arduino-ESP32.

**Reply by jwestes** [reply]
I just pulled the latest code from the master branch, and if I wrap EVE_target.h in an extern "C", I get all sorts of errors.  As soon as I move that one line out and go back to the includes I posted 15 minutes ago, they go away and everything works.

**Reply by RudolphRrr** [reply]
I did not try to compile for Arduino-ESP32 earlier.
I just did now and at first it did not compile - as expected - then I copied over the modified EVE.h and now it compiles:

Processing ESP32 (platform: espressif32; board: esp32doit-devkit-v1; framework: arduino)
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------Verbose mode can be enabled via `-v, --verbose` option
CONFIGURATION: https://docs.platformio.org/page/boards/espressif32/esp32doit-devkit-v1.html
PLATFORM: Espressif 32 (6.0.1) > DOIT ESP32 DEVKIT V1
HARDWARE: ESP32 240MHz, 320KB RAM, 4MB Flash
DEBUG: Current (cmsis-dap) External (cmsis-dap, esp-bridge, esp-prog, iot-bus-jtag, jlink, minimodule, olimex-arm-usb-ocd, olimex-arm-usb-ocd-h, olimex-arm-usb-tiny-h, olimex-jtag-tiny, tumpa)
PACKAGES:
 - framework-arduinoespressif32 @ 3.20006.221224 (2.0.6)
 - tool-esptoolpy @ 1.40400.0 (4.4.0)
 - toolchain-xtensa-esp32 @ 8.4.0+2021r2-patch5
LDF: Library Dependency Finder -> https://bit.ly/configure-pio-ldf
LDF Modes: Finder ~ chain, Compatibility ~ soft
Found 34 compatible libraries
Scanning dependencies...
Dependency Graph
|-- EmbeddedVideoEngine @ 5.0.6
|   |-- SPI @ 2.0.0
|-- SPI @ 2.0.0
Building in release mode
Compiling .pio\build\ESP32\src\EVE_Test.cpp.o
Compiling .pio\build\ESP32\src\tft.c.o
Compiling .pio\build\ESP32\lib272\EmbeddedVideoEngine\EVE_commands.c.o
Linking .pio\build\ESP32\firmware.elf
Retrieving maximum program size .pio\build\ESP32\firmware.elf
Checking size .pio\build\ESP32\firmware.elf
Advanced Memory Usage is available via "PlatformIO Home > Project Inspect"
RAM:   [=         ]   8.1% (used 26400 bytes from 327680 bytes)
Flash: [==        ]  20.3% (used 265569 bytes from 1310720 bytes)
Building .pio\build\ESP32\firmware.bin
esptool.py v4.4
Creating esp32 image...
Merged 2 ELF sections
Successfully created esp32 image.
=============================================================================== [SUCCESS] Took 5.21 seconds ===============================================================================
Environment    Status    Duration
-------------  --------  ------------
ESP32          SUCCESS   00:00:05.209

My EVE_Test.cpp is this currently:
`
#include <Arduino.h>
#include <SPI.h>
#include "EVE.h"
#include "tft.h"
#include "tft_data.h"

void setup()
{
    pinMode(EVE_CS, OUTPUT);
    digitalWrite(EVE_CS, HIGH);
    pinMode(EVE_PDN, OUTPUT);
    digitalWrite(EVE_PDN, LOW);

/* we are not using the Arduino SPI class since we want DMA */
#if defined(ESP32) || (ARDUINO_NUCLEO_F446RE) || (WIZIOPICO) || (PICOPI)
    EVE_init_spi();
#else
    SPI.begin(); /* sets up the SPI to run in Mode 0 and 1 MHz */
    SPI.beginTransaction(SPISettings(8000000, MSBFIRST, SPI_MODE0));
#endif

    EVE_init();

    EVE_cmd_inflate(0, logo, sizeof(logo)); /* load logo into gfx-memory and de-compress it */
    EVE_cmd_loadimage(1000, EVE_OPT_NODL, pic, sizeof(pic));

//    TFT_init();
    // SPI.beginTransaction(SPISettings(20000000, MSBFIRST, SPI_MODE0));

#if defined(ESP8266)
    SPI.setFrequency(16000000);
#endif
}

void loop()
{
    uint32_t current_millis;
    static uint32_t previous_millis = 0;
    static uint8_t display_delay = 0;
    uint32_t micros_start, micros_end;
    current_millis = millis();

    EVE_memWrite32(REG_TOUCH_TRANSFORM_A, 0x000074df);
    EVE_start_cmd_burst();
    EVE_cmd_fgcolor_burst(0x00c0c0c0UL);
    EVE_cmd_setbitmap_burst(0, EVE_RGB565, 100U, 100U);

    if ((current_millis - previous_millis) > 4) /* execute the code every 5 milli-seconds */
    {
        previous_millis = current_millis;

        micros_start = micros();
        TFT_touch();
        micros_end = micros();
        num_profile_b = (micros_end - micros_start); /* calculate the micro-seconds passed during the call to TFT_touch */

        display_delay++;
        if (display_delay > 3) /* refresh the display every 20ms */
        {
            display_delay = 0;
            micros_start = micros();
            TFT_display();
            micros_end = micros();
            num_profile_a = (micros_end - micros_start); /* calculate the micro-seconds passed during the call to TFT_display */
        }
    }
}
`

So this is not supposed to be working, this is only to compile.

**Reply by RudolphRrr** [reply]
This is the EVE.h I am using for a quick test.
[EVE.h.zip](https://github.com/RudolphRiedel/FT800-FT813/files/11119678/EVE.h.zip)

**Comment by jwestes** [comment]
Using that file, this is an example of the error messages I'm seeing.  I get hundreds of them, so I will just show you the first one.

Using previously compiled file: C:\Users\jeste\AppData\Local\Temp\VMBuilds\CompleteSurfSystem.Controller\esp32_esp32\Debug 7-in\Gyroscope.cpp.o
arduino.h: 143:6: error: conflicting declaration of C function 'long int random(long int)
... LOTS of compler crap goes here...
   long random(long)
   ^~~~~~

This header file works for me though.
[EVE.zip](https://github.com/RudolphRiedel/FT800-FT813/files/11119771/EVE.zip)

**Reply by RudolphRrr** [reply]
I do not see why this would make any difference since EVE_target.h includes EVE_target_Arduino_ESP32.h and this is also wrapped in #ifdef __cplusplus / extern "C".
I can compile with your version as well.

But as the error above states "previously compiled" - does this happen with clean / rebuild as well?

**Comment by jwestes** [comment]
I just started a rebuild to see.  It takes about 15 minutes and it is 2am here, so I am going to bed, but I will let you know in the morning what I have found.

**Reply by jwestes** [reply]
A clean / rebuild did not help.

**Reply by RudolphRiedel** [reply]
Ok, I implemented the change in EVE.h with the #include "EVE_target.h" line left out.
I tested it and while I still would like to know why this is not working the other way around for you,
I found no downside to implementing it this way either.
My Arduinio example builds just fine and I have it up and running on a Teensy 4.1 right now with an extra line in setup().
And the plain C examples are not affected either.

The only issue I found while testing is that my ESP32 example does not build anymore.
There was an update to "Espressif 32" platform for PlatformIO and now the first error I get is:
src/main.c:12:10: fatal error: soc/rtc_wdt.h: No such file or directory
Totally unrelated - but what the heck? :-)

**Reply by jwestes** [reply]
I just pulled the latest library and I now have the ability to use it as-is with no compiler issues.  Thank you!

**Reply by RudolphRiedel** [reply]
Nice.
The ESP32 issue is a result of that the PlatformIO "Espressif 32" platform was updated to use ESP-IDF 5.0.
Espressif did make a couple of breaking changes from 4.x to 5.x and what I ran into is not even documented as such.
This is the ESP-IDF version, the Arduino ESP32 version might not be affected.

**Comment by jwestes** [comment]
I don't use PlatformIO, so I have not seen this.  I'm using Visual Studio with VisualMicro.

**Reply by RudolphRiedel** [reply]
PlatformIO is not the issue ESP-IDF 5.x is. 
Or was, when I am done testing it.

Ah, Arduino-ESP32 2.0.7 is still using ESP-IDF 4.4.4.

**Comment by Brechtjeu** [comment]
Sorry for the delay.

Personally I compile with the latest platformio for arduino-esp32 (`platformIO Core, version 6.1.6`)

I have gotten it to work by removing all `extern "c"` e.g. from tft.h and and changing `tft.c` to `tft.cpp`. As far as I am aware i did not change anything else, and I can now compile with all cpp files except for tft_data.c (its the only thing that gives me errors when I try compiling it with a cpp extension)

**Reply by RudolphRrr** [reply]
I just downloaded the archive, grabbed the EVE_Test_Arduino_PlatformIO example and compiled it for ESP32 using pio 6.1.6 and Espressif 32 in version 6.0.1 and 6.1.0 - it just compiles as is and for every other platform in that example.

Compiling this way where a C++ part is combined with a plain C part is working fine for years now.
The only tricky part to solve was to call C++ class methods from plain C functions.
What really did not work was calling the functions from C++.

Please provide a minimal platformio project that does not compile and I'll have a look.
